File: source/python/lcs.py
def lcs(a, b):
    m, n = len(a), len(b)
    dp = [[0] * (n + 1) for _ in range(m + 1)]
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if a[i - 1] == b[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

    result = ""
    i, j = m, n
    while i > 0 and j > 0:
        if a[i - 1] == b[j - 1]:
            result = a[i - 1] + result
            i -= 1
            j -= 1
        elif dp[i - 1][j] >= dp[i][j - 1]:
            i -= 1
        else:
            j -= 1

    return {"length": dp[m][n], "subsequence": result}

a = "ABCBDAB"
b = "BDCAB"
result = lcs(a, b)
print(f"lcs length: {result['length']}")
print(f"lcs: {result['subsequence']}")

File: source/javascript/lcs.js
// Longest Common Subsequence (LCS)
function lcs(a, b) {
    const m = a.length, n = b.length;
    const dp = Array.from({ length: m + 1 }, () => new Array(n + 1).fill(0));

    for (let i = 1; i <= m; i++) {
        for (let j = 1; j <= n; j++) {
            if (a[i - 1] === b[j - 1]) dp[i][j] = dp[i - 1][j - 1] + 1;
            else dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
        }
    }

    let result = "";
    let i = m, j = n;
    while (i > 0 && j > 0) {
        if (a[i - 1] === b[j - 1]) { result = a[i - 1] + result; i--; j--; }
        else if (dp[i - 1][j] >= dp[i][j - 1]) i--;
        else j--;
    }

    return { length: dp[m][n], subsequence: result };
}

const a = "ABCBDAB", b = "BDCAB";
const result = lcs(a, b);
console.log(`lcs length: ${result.length}`);
console.log(`lcs: ${result.subsequence}`);
